### PR TITLE
Compactor: Add tracing support

### LINKED ISSUE #4832: Add tracing support for compaction
### Is your proposal related to a problem?

We have some metrics in the compactor of its overall status. But right now it is hard to track the status and time taken for each compaction run.

### Describe the solution you'd like

Add tracing support for compactor and instrument different steps in compaction. For example:
1. Block planning
2. Block download
3. Block verification
4. Block compaction
5. Block upload

### PR DESCRIPTION
**Pull request by metonymic-smokey**
Fixes #4832

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
Added tracing support for the various compaction steps(similar to Querier). 

## Verification
Tested locally using Jaeger.

### CODE REVIEW

File: pkg/compact/compact.go
Comment (matej-g) [inline pkg/compact/compact.go:1029]: Nit: Maybe `_healthstats` since the method does not check, only collects stats?
```suggestion
		tracing.DoInSpan(ctx, "compaction_block_healthstats", func(ctx context.Context) {
```

**Review by matej-g — COMMENTED** [review]
This is looking cool! I was wondering about one more thing, which I think could be particularly useful in the context of compaction and that is to also record the error in the span (we would have the error in the logs, but I think having them directly on span would be more handy). 

We are using our convenience method `DoInSpan` here, but that does not capture an error if it occurs. Maybe we could have another method (e.g. `DoInSpanWithErr`) which could also set the error on span. Does that sound good?

File: pkg/compact/compact.go
Comment (metonymic-smokey) [reply pkg/compact/compact.go:1029]: Logging errors sounds good! I had tried it one of the earlier commits but reverted considering errors were already being logged. 
Instead of a custom function, there is an existing function, `span.LogFields(opentracing_log.Error(err))`. Maybe I could use that?

Comment (matej-g) [reply pkg/compact/compact.go]: I think the issue is that if we use our "custom" method `DoInSpan`, we do not have access to the actual span. Hence why I suggested a kind of enhanced version of the same method. Other option would be to not use `DoInSpan` and instrument the code "manually".

Comment (metonymic-smokey) [reply pkg/compact/compact.go]: Thanks.
There is a way to get the span from the current context, which I have used in my recent commit. That way, I don't have to instrument manually and can log errors with minimal changes.

Comment (metonymic-smokey) [reply pkg/compact/compact.go]: I have pushed a commit with an implementation of the former. Please let me know what you think.
thanks!

File: pkg/tracing/tracing.go
Comment (matej-g) [inline pkg/tracing/tracing.go:83]: 1) I'm not sure about the behavior if `err` is `nil` (whether it still logs error), but maybe we could check `err != nil` before we log error?
2) I think we can replace this with method from the `ext` package, see https://pkg.go.dev/github.com/opentracing/opentracing-go@v1.2.0/ext#LogError. This will set all necessary things on the span for us

**Review by matej-g — COMMENTED** [review]
Great job @metonymic-smokey! Thanks for going above and beyond to address things. I'm wondering about one last thing, afterwards I believe this is ready.

Let's try to get a review from maintainer as well, perhaps @yeya24 as original issue reporter.

File: pkg/tracing/tracing.go
Comment (metonymic-smokey) [reply pkg/tracing/tracing.go]: Good point @matej-g. Thanks!
I've made the changes in a recent commit. Let me know if they're okay.

File: pkg/tracing/tracing.go
Comment (matej-g) [inline pkg/tracing/tracing.go:85]: I believe we do not need to pass the last value, since the method will already include the field for error.

**Review by matej-g — APPROVED** [review]
:100:

File: pkg/compact/compact.go
Comment (squat) [inline pkg/compact/compact.go:1025]: tiny nit
```suggestion
```

File: pkg/compact/compact.go
Comment (squat) [inline pkg/compact/compact.go:1032]: Since the Golang name is `...HealthStats`, `health` and `stats` should be treated as separate words
```suggestion
		tracing.DoInSpanWithErr(ctx, "compaction_block_health_stats", func(ctx context.Context) error {
```

File: pkg/compact/compact.go
Comment (squat) [inline pkg/compact/compact.go:770]: All other spans are named `compaction_<x>`, should we make this follow the same convention?

**Review by squat — APPROVED** [review]
Looks good overall! Some very small nits

**Review by squat — APPROVED** [review]

**Comment by squat** [comment]
lgtm! restarting failed jobs and we can merge :))

**Comment by metonymic-smokey** [comment]
Thanks for the reviews @matej-g @squat !

**Pull request merged**
